File: client/go/src/example/event/event.go
package event

import (
	"fmt"
	"context"
	"biostar/service/event"
	"google.golang.org/grpc"
)

type EventSvc struct {
	client event.EventClient
}

func NewEventSvc(conn *grpc.ClientConn) *EventSvc {
	return &EventSvc{
		client: event.NewEventClient(conn),
	}
}

func (s *EventSvc) GetLog(deviceID, startEventID, maxNumOfLog uint32) ([]*event.EventLog, error) {
	req := &event.GetLogRequest{
		DeviceID: deviceID,
		StartEventID: startEventID, 
		MaxNumOfLog: maxNumOfLog,
	}

	resp, err := s.client.GetLog(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot get the log events: %v\n", err)

		return nil, err
	}

	return resp.GetEvents(), nil
}


func (s *EventSvc) GetImageLog(deviceID, startEventID, maxNumOfLog uint32) ([]*event.ImageLog, error) {
	req := &event.GetImageLogRequest{
		DeviceID: deviceID,
		StartEventID: startEventID, 
		MaxNumOfLog: maxNumOfLog,
	}

	resp, err := s.client.GetImageLog(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot get the image logs: %v\n", err)

		return nil, err
	}

	return resp.GetImageEvents(), nil
}


const (
	MONITORING_QUEUE_SIZE = 8
)

var (
	eventStream event.Event_SubscribeRealtimeLogClient
)

func (s *EventSvc) StartMonitoring(deviceID uint32) (context.CancelFunc, error) {
	enableReq := &event.EnableMonitoringRequest{
		DeviceID: deviceID,
	}

	_, err := s.client.EnableMonitoring(context.Background(), enableReq)

	if err != nil {
		fmt.Printf("Cannot enable log monitoring: %v\n", err)
		return nil, err
	}

	subReq := &event.SubscribeRealtimeLogRequest{
		QueueSize: MONITORING_QUEUE_SIZE,
		DeviceIDs: []uint32 { deviceID },
	}

	ctx, cancelFunc := context.WithCancel(context.Background())

	eventStream, err = s.client.SubscribeRealtimeLog(ctx, subReq)

	if err != nil {
		fmt.Printf("Cannot enable log monitoring: %v\n", err)
		return nil, err
	}

	go func() {
		fmt.Printf("Start receiving real-time events\n")

		for {
			eventLog, err := eventStream.Recv()

			if err != nil {
				fmt.Printf("Cannot receive real-time events: %v\n", err)
				return
			}

			fmt.Printf("Event: %v\n", eventLog)
		}

	} ()

	return cancelFunc, nil
}


func (s *EventSvc) StopMonitoring(deviceID uint32) error {
	disableReq := &event.DisableMonitoringRequest{
		DeviceID: deviceID,
	}

	_, err := s.client.DisableMonitoring(context.Background(), disableReq)

	if err != nil {
		fmt.Printf("Cannot disable log monitoring: %v\n", err)
		return err
	}

	fmt.Printf("Stop receiving real-time events\n")

	return nil
}

File: client/python/example/event/event.py
import grpc

import event_pb2_grpc
import event_pb2


class EventSvc:
  stub = None

  def __init__(self, channel): 
    try:
      self.stub = event_pb2_grpc.EventStub(channel)
    except grpc.RpcError as e:
      print(f'Cannot get the event stub: {e}')
      raise

  def getLog(self, deviceID, startEventID, maxNumOfLog):
    try:
      response = self.stub.GetLog(event_pb2.GetLogRequest(deviceID=deviceID, startEventID=startEventID, maxNumOfLog=maxNumOfLog))
      return response.events
    except grpc.RpcError as e:
      print(f'Cannot get the event log: {e}')
      raise

  def getImageLog(self, deviceID, startEventID, maxNumOfLog):
    try:
      response = self.stub.GetImageLog(event_pb2.GetImageLogRequest(deviceID=deviceID, startEventID=startEventID, maxNumOfLog=maxNumOfLog))
      return response.imageEvents
    except grpc.RpcError as e:
      print(f'Cannot get the image events: {e}')
      raise    

  def enableMonitoring(self, deviceID):
    try:
      self.stub.EnableMonitoring(event_pb2.EnableMonitoringRequest(deviceID=deviceID))
    except grpc.RpcError as e:
      print(f'Cannot enable monitoring: {e}')
      raise

  def disableMonitoring(self, deviceID):
    try:
      self.stub.DisableMonitoring(event_pb2.DisableMonitoringRequest(deviceID=deviceID))
    except grpc.RpcError as e:
      print(f'Cannot disable monitoring: {e}')
      raise

  def subscribe(self, queueSize): 
    try:
      return self.stub.SubscribeRealtimeLog(event_pb2.SubscribeRealtimeLogRequest(queueSize=queueSize))
    except grpc.RpcError as e:
      print(f'Cannot subscribe: {e}')
      raise
